Entrega 1

Natygimep opened the pull request:
No está como quisiéramos pero hay que cumplir la fecha de entrega.

Natygimep committed:
Inicié develop

Natygimep committed:
Tímido inicio, métodos iniciales.

Natygimep committed:
Tratando de acomodar los objetos. Está verde.

Commit:
Agregue objetos y incorpore metodos a joaquin y lucia

Objetos agregados Joaquin, lucia y luisAlberto (todvia no lo toque) y las dos canciones cisne y la familia

Commit:
Objetos nuevos + método en objeto luisAlberto

Natygimep committed:
Completé letra de canciones.

Natygimep committed:
Cambié el método valor de fender por una constante y agregué la fender

Commit:
Nuevo método en Objeto joaquin y lucia 

Método nuevo en joaquin habilidad (si canta solo o acompañado) + como cobra.
Método nuevo en lucia como cobra.

Commit:
Cambios + Arreglos + Nuevo método

Arreglos en el método habilidad de lucia y algunas cosas turbias :)

Commit:
Nuevo método objeto luisAlberto

Saque todo lo de objeto mundoMusical.

Commit:
Arrelgos

Commit:
Arreglo chiquito

braupalizas committed:
Un pequeño arreglo en algunas cositas

Natygimep committed:
Agregué todos los test pedidos, pero ello no significa que funcionen.

Natygimep committed:
Cambié la habilidad de Joaquín

Natygimep committed:
Modifiqué la habilidad de Lucía

Natygimep committed:
Los test de canciones de Joaquín y Lucía andan bien.

Natygimep committed:
Los test de lo que cobra la hora Joaquín andan bien

Natygimep committed:
Los test de lo que cobra Lucía funcionan

Natygimep committed:
Modifiqué los métodos de Gibson para la habilidad de Luis Alberto.

Natygimep committed:
Los test de Luis Alberto funcionan bien.

Commit:
Agregue nuevos temas + dos musicos

braupalizas committed:
Agrego clases y un par de retoques

Commit:
Nuevo métodos en las Canciones +  Nuevos métdos en Luis Alberto

Puede ser que este todo hiperhardkodeado jaj pero funciona, Falta el tema de las letras minúsculas y mayúsculas que no lo encontre y el método del musico que si la "pégo" o no.

sanuzzi commented on src/mundoMusical.wlk:
```diff
@@ -0,0 +1,236 @@
+		return precio*horas
+	}
+	
+	method tocaEnGrupo(aumentoHabilidad) {
```
Que le digamos a un músico de grupo, que toque en grupo, es medio raro...

sanuzzi commented on src/mundoMusical.wlk:
```diff
@@ -0,0 +1,236 @@
+		}
+	}
+	
+	method tocaEnGrupo() {
```
Que le digamos a un solista, que toque en grupo, también es raro...

sanuzzi commented on src/mundoMusical.wlk:
```diff
@@ -0,0 +1,236 @@
+	
+	method interpretaBien(cancion) = (cancion.letra().contains(palabra))
+	
+	method cobra(lugar) {
```
Pueden poner directamente:
method cobra(lugar) **= if (condición) valor1 else valor2**

sanuzzi commented on src/mundoMusical.wlk:
```diff
@@ -0,0 +1,236 @@
+	
+	method todasLasCanciones() = albumParaLosArboles + albumJustCrisantemo 
+	
+	method esMinimalista() = self.todasLasCanciones().all({cancion => cancion.duracion() < 180})
```
**Todos** los músicos deben entender el mensaje "sosMinimalista". No solo Luis.

sanuzzi commented on src/mundoMusical.wlk:
```diff
@@ -0,0 +1,236 @@
+	
+	method esMinimalista() = self.todasLasCanciones().all({cancion => cancion.duracion() < 180})
+	
+	method cancionMasLargaDeAlbumParaLosArboles() = self.albumParaLosArboles().max({cancion => cancion.longitudDeLetra()}) // si lo se que es feo, pero bueno funciona jeje
```
Hay que conocer cuál es la canción más larga de un **álbum**. Así que los **álbumes** deben tener este método.

Ustedes hicieron que un álbum sea una lista, y conviene que sea un objeto de una clase nueva hecha por ustedes. Y que adentro, tenga una referencia al título del álbum, canciones, y lo que necesite.

sanuzzi commented on src/mundoMusical.wlk:
```diff
@@ -0,0 +1,236 @@
+	
+	method apareceEn(palabra) = self.todasLasCanciones().filter({cancion => cancion.letra().contains(palabra)}) // Este m�todo tiene problemas con las letras May�sculas y min�sculas 
+	
+	method habilidad(guitarra) = [100,guitarra.valor() * 8].min()   // minimo entre 100 y el valor x 8 ya que el tope es 100
```
los números entienden min:

5.min(7) --> 5
7.min(5) --> 5

y además: Todo músico tiene que entender un mismo mensaje para decirnos su habilidad.

Es raro que a unos le pasemos una guitarra y a otros no.

Hagan que todos entiendan lo mismo. En cuyo caso, necesitarían indicarle a Luis cuál es su guitarra actual, que se la **guarde**, y al preguntarle la habilidad, que ahí le pregunte a **su** guitarra cuál es su valor. ¿Me explico?

sanuzzi commented on src/mundoMusical.wlk:
```diff
@@ -0,0 +1,236 @@
+	
+}
+
+object joaquin inherits MusicoDeGrupo{}
```
Créenlo en el **fixture**. Nada de crear un WKO solo para un flaco más de muchos que podríamos tener en el programa.

Lo mismo para **todos** los músicos.

sanuzzi commented on src/mundoMusical.wlk:
```diff
@@ -0,0 +1,236 @@
+	 	valor = nuevoValor
+	 }      
+	
+	method maltrataGibson() {
```
¿No se puede arreglar?

Además, estaría bueno que se guarden un booleano para saber si está rota. Y en base a eso, devolver un valor u otro directamente. No hace falta la referencia "valor".

sanuzzi commented on src/mundoMusical.wlk:
```diff
@@ -0,0 +1,236 @@
+	
+	method duracion() = 145
+	
+	method longitudDeLetra() = self.letra().size()
```
Repiten en todas las canciones ---> Hagan una **clase**!!!

Además, las canciones van a necesitar decirnos más cosas que ésas, como lo de saber si es **corta** (menos de 3 minutos).

sanuzzi commented on src/test.wtest:
```diff
@@ -0,0 +1,105 @@
+import mundoMusical.*
+
+describe "tests respecto de los m�sicos" {
```
Hagan 2 archivos de tests. Uno para cada entrega.

Y asegúrense de que **todos** queden andando con el nuevo código. Les faltan tests!!

Por último, usen un **fixture** en cada archivo para crear a los objetos que deban instanciarse de clases

sanuzzi commented on src/mundoMusical.wlk:
```diff
@@ -0,0 +1,236 @@
+class MusicoDeGrupo {
```
El músico tiene que saber hacer más cosas:

Decirnos si es minimalista, qué canciones tienen cierta palabra, si la pegó, etc.

Ese comportamiento en común (y quizá más cosas), deben estar en una **única clase** "Músico".

Luego, podrían tener 2 subclases: MúsicoDeGrupo y VocalistaPopular (o MúsicoSolista, como dijeron ustedes).

El problema de eso, es que cada músico puede pasar a ser solista o tener grupo, y es raro que alguien de clase MúsicoSolista, tenga grupo...

Por lo tanto, **en vez de usar herencia**, ¿Qué se les ocurre?

Con esto se anticipan a la entrega 4, que les pide explícitamente que alguien cambie su tipo de músico.

sanuzzi commented on src/test.wtest:
```diff
@@ -0,0 +1,105 @@
+
+/*test "La habilidad de Joaqu�n cantando en Pimpinela debe ser 25" {
+
+	assert.equals(joaquin.habilidad("conLucia"), 25)
```
No le pasen un string.

sanuzzi reviewed (changes requested):
Revisión entregas 1 y 2

sanuzzi commented on src/mundoMusical.wlk:
```diff
@@ -0,0 +1,236 @@
+	
+	method interpretaBien(cancion) = (cancion.duracion() > 300)
+	
+	method cobra(horas) {
```
No hay que hacer esto. Directamente queremos saber su precio por hora y ya.

Y ojo porque un músico de grupo cobra diferente en base a si está él solo en la **presentación**.

Así que sí o sí tenemos que recibir la **presentacióon** para la cual nos piden el costo, y adentro del método, preguntarle a la presentación si solo actúa "yo". ¿Me explico?

sanuzzi reviewed (changes requested):
Comentario adicional de entregas 1 y 2

Natygimep committed:
Tests de la primer entrega

Acomodé el código para que los test den verde, faltan algunos.

Natygimep committed:
Mas tests

Faltan los test con el manejo de fechas que no me sale.

Natygimep committed:
Agregué código comentado

Son los cambios de la segunda entrega que saqué para redondear la primer entrega.

Natygimep committed:
Capacidad trastienda funciona

Natygimep committed:
Fin entrega uno

Todos los métodos con fechas funcionan.

Natygimep committed:
Descripción de los test

Falta poner el código pero adelanté algo de lo que hay que escribir y que es mucho.

Natygimep committed:
Separando archivos

Guardo hasta donde llegué, sigo mañana.

Natygimep committed:
encaré las clases nuevas

Todavía falta pero lo subo por las dudas para no perderlo.

Natygimep committed:
Armé las clases nuevas.

Todavia no funcionan todos los tests de la entrega 1.

Natygimep committed:
Tratando de parametrizar interpreta bien y cobra.

Natygimep committed:
Pasé las canciones a fixture de los test

braupalizas committed:
Armamos los setters

braupalizas committed:
Creando las subclases musicoGrupo y musicoSolista

braupalizas committed:
Haciendo funcionar de nuevo tests entrega 1

braupalizas committed:
Tests 1 con luis alberto funcionando

Natygimep committed:
Agregué el método esCorta en la clase Canción

Natygimep committed:
Completé la clase Album y los di de alta en fixture.

sanuzzi commented on src/Album.wlk:
```diff
@@ -0,0 +1,40 @@
+class Album {
+	var titulo
```
¿Por qué todo variable? ¿Tiene sentido que cambie con el tiempo?
Revisen en todas sus clases.

Natygimep replied:
Corregido.

sanuzzi commented on src/Album.wlk:
```diff
@@ -0,0 +1,40 @@
+	var titulo
+	var canciones
+	var artista
+	var copiasEditadas
```
¿Editadas? ¿Para qué?

Natygimep replied:
Las copias editadas son las copias lanzadas, que usamos para calcular si la pegó.

sanuzzi commented on src/canciones.wlk:
```diff
@@ -0,0 +1,28 @@
+method titulo(unTitulo){
+	titulo = unTitulo
+}
+method esCorta(unaCancion){
```
Pueden usar "= cosaQueHayQueDevolver" en vez del return. Revisen en otros lugares.

Natygimep replied:
Corregido.

sanuzzi commented on src/mundoMusical.wlk:
```diff
@@ -0,0 +1,152 @@
+	
+}
+
+class MusicoDeGrupo inherits Musico {
```
Si queremos que un músico de grupo pase a ser solista, no puede.

Entonces, no elijan herencia. Repensar esta parte.

Natygimep replied:
Corregido.

sanuzzi commented on src/mundoMusical.wlk:
```diff
@@ -0,0 +1,152 @@
+	
+}
+
+object luisAlberto inherits Musico{
```
No hace falta un WKO para Luis. Que sea un Músico normal.

Una vez que repiensen lo de estar en grupo o ser solista, van a ver que lo único que cambia de Luis respecto a eso, es justamente esa parte. Serían 3 posibilidades: Estar en grupo, ser solista, o ser el tipo de músico de Luis Alberto (o tipo de músico re capo, como lo quieran llamar).

Entonces, no hace falta ni que Luis tenga una clase diferente ni que sea un WKO, pero sí que tenga un tipo de músico especial para él.

sanuzzi commented on src/mundoMusical.wlk:
```diff
@@ -0,0 +1,152 @@
+	}
+	method interpretaBien(cancion) = true
+	
+	method cobra(fecha){
```
Sería más corto:
 = if(fecha >= new Date(1,9,2017) ) 1200 else 1000

Natygimep replied:
Corregido.

sanuzzi commented on src/mundoMusical.wlk:
```diff
@@ -0,0 +1,152 @@
+		}
+	}
+	
+	override method habilidad(){
```
Las colecciones entienden min, pero también los **números**.
No hace falta hacer una colección intermedia.

Un número entiende "min", recibiendo otro número, y devuelve el más pequeño de los 2.

Natygimep replied:
Corregido.

sanuzzi commented on src/mundoMusical.wlk:
```diff
@@ -0,0 +1,152 @@
+		capacidad = unaCapacidad
+	}
+	
+	method capacidad() = return capacidad
```
Ojo con returns al vicio

Natygimep replied:
Corregido. PERDONNNNNNNNNNNNNNN

sanuzzi commented on src/mundoMusical.wlk:
```diff
@@ -0,0 +1,152 @@
+
+// Lugares donde tocan los m�sicos 
+
+class Lugar {
```
Es medio raro que puedan instanciar esta clase para el luna park y no para la trastienda. Quizá convenga decidirse por una opción.

Quizá ambos pueden ser wko, y que tengan una superclase "Lugar" con 2 cosas:
1. Un método abstracto "capacidadEn(unaFecha)"
2. lo de saber si es concurrido.

Luego, el wko del luna, implementará lo de capacidad, y directamente podría devolver el número que corresponda, sin necesitar un constructor ni nada.

Y la trastienda, tampoco necesita guardarse la capacidad base en una referencia.
Si la fecha cae Sábado, es tanto, y si no, tanto otro.
O bien, sería la capacidad base + capacidad adicional si es sábado (y esto último devuelve el extra o 0 según el día).
Todos los métodos pueden devolver valores directamente, sin depender de referencias.

sanuzzi commented on src/mundoMusical.wlk:
```diff
@@ -0,0 +1,152 @@
+
+
+/*const albumParaLosArboles = [cisne,almaDeDiamante]
+	const albumJustCrisantemo = [crisantemo]
```
¿Y esto?

Natygimep replied:
Lo que estaba comentado?

sanuzzi commented on src/test.wtest:
```diff
@@ -0,0 +1,142 @@
+    joaquin = new MusicoDeGrupo(20, 5)
+	fender = new Guitarra(10)
+	gibson = new Guitarra(15)
+	cisne = new Cancion("Hoy el viento se abri� qued� vac�o el aire una vez m�s y el manantial brot� y nadie est� aqu� y puedo ver que solo estallan las hojas al brillar", 312,"cisne")
```
Conviene que los argumentos estén en este orden:
1. Título
2. Duración
3. Letra

sanuzzi commented on src/test.wtest:
```diff
@@ -0,0 +1,142 @@
+test "El costo de la presentaci�n de los tres en el La Trastienda el 15/11/2017 es 1650" {
+	joaquin.tocaEnGrupo()
+	const unaFecha = new wollok.lang.Date(15,11,2017)
+	assert.equals(lucia.cobra(laTrastienda)+joaquin.cobra(laTrastienda)+luisAlberto.cobra(unaFecha), 1650)
```
La presentación debería saber el costo total de los músicos que se presentan en ella

sanuzzi commented on src/testEntrega2.wtest:
```diff
@@ -0,0 +1,88 @@
+    luisAlberto = new Musico()
+    sole = new Musico(55)
+    kike = new Musico(60)
+    cisne = new Cancion("Hoy el viento se abri� qued� vac�o el aire una vez m�s y el manantial brot� y nadie est� aqu� y puedo ver que solo estallan las hojas al brillar", 312, "Cisne")
```
Para evitar repetir, podrían hacer un objeto auxiliar que tenga:

1. Las constantes:
const cisne = new Cancion( ... )

2. Los getters de esas constantes.

3. Un método como "inicializate" para hacer este tipo de tareas:
luisAlberto.albumes( [ paraLosArboles, justCrisantemo ] )

De esta forma, su fixture les quedaría así:

fixtureAuxiliar.inicializate()

cisne = fixtureAuxiliar.cisne()
laFamilia = fixtureAuxiliar.laFamilia()
etc.

De esta forma, no se ahorran "todo", pero al menos no instancian lo mismo una y otra vez en cada archivo de tests.

sanuzzi commented on src/testEntrega2.wtest:
```diff
@@ -0,0 +1,88 @@
+    albumJoaquin = new Album("Especial la familia", kike, [laFamilia], 100000, 89000, 17/06/1992)
+    albumSole = new Album("La Sole", sole, [eres, corazonAmericano], 200000, 130000, 04/02/2005)
+    albumLuisAlberto1 = new Album("Para los arboles", luisAlberto, [cisne, almaDeDiamante], 50000, 49000, 31/03/2003)
+    albumLuisAlberto2 = new Album("Just crisantemo", luisAlberto, [crisantemo], 28000, 27500, 05/12/2007)
```
Si el álbum se llama Just crisantemo, la referencia podría ser justCrisantemo.

Los tests de más abajo se beneficiarían de esa expresividad.

Además, hace falta la parte en donde un artista conoce a sus álbumes, están inconexos hasta acá

Natygimep replied:
Corregido.

sanuzzi commented on src/testEntrega2.wtest:
```diff
@@ -0,0 +1,88 @@
+	assert.equals(false)
+}
+test "Joaqu�n la peg�"{
+	assert.equals(false)
```
Falta terminar, y falta entrega 3 completa

sanuzzi reviewed (commented):
Revisión entrega 3

Natygimep committed:
Agregué el método canciones de un artista

Natygimep committed:
Cambiamos las variables por constantes.

Desde acá hasta nuevo aviso, estamos reunidos en grupo usando mi usuario.

Natygimep committed:
Pasamos var a const en las clases que faltaban.

Natygimep committed:
Cambiamos herencia por composición en Músico.

Natygimep committed:
Creamos la clase presentacion, y mas modificaciones

Natygimep committed:
No anda el test duracion obra.

Fin trabajo grupal.

Commit:
Revert "Auto stash before merge of "develop" and "origin/develop""

This reverts commit 8925d12032153fd59d48778f83318cc7070677dc.

Commit:
Auto stash before merge of "develop" and "origin/develop"

braupalizas committed:
Completando duracionDeObra , esMinimalista y un par de detalles mas

braupalizas committed:
Sumamos canciones con determinada palabra

braupalizas committed:
Finalizando entrega 2

Commit:
Acomodo el codigo para que se vea mas bonito!

braupalizas committed:
Haciendo volver a funcionar los tests 1 y 2

braupalizas committed:
Encarando entrega 3

braupalizas committed:
Creando objeto auxiliar para el fixture

sanuzzi commented on src/Album.wlk:
```diff
@@ -43,12 +43,26 @@ class Album {
+}
+
+object compararPorLetra {
+	method esLaMayor(unaCancion) = unaCancion.letra().size()
```
Mal nombre. No quieren saber si esLaMayor, sino el **tamaño** en base a lo que considere ese criterio

sanuzzi commented on src/canciones.wlk:
```diff
@@ -8,7 +8,7 @@ class Cancion {
 	const artista
 
-	constructor(unaLetra, unaDuracion, unTitulo, unArtista){
+	constructor(unTitulo, unaDuracion, unArtista, unaLetra){
```
¿Para qué quieren al artista?

sanuzzi commented on src/guitarra.wlk:
```diff
@@ -8,5 +8,9 @@ class Guitarra {
-
+	
+	method maltratar() {
+		self.valor(5)
```
No asignen nuevos valores numéricos ya que es más difícil de mantener.
Cambien un booleano y nada más.
En el método para devolver el valor, digan si está rota o no, y en base a eso devuelvan cierto número fijo u otro.

sanuzzi commented on src/mundoMusical.wlk:
```diff
@@ -1,41 +1,41 @@
-	var habilidad
 	var tipoDeMusico
-	const albumes
+	const albumes = []
```
¿Por qué un lista?

sanuzzi commented on src/mundoMusical.wlk:
```diff
@@ -1,41 +1,41 @@
 	method canciones() = self.albumes().flatMap({unAlbum => unAlbum.canciones()})
 	
-	method agregarAlbum(unAlbum) = self.albumes().add(unAlbum)	
+	method agregarAlbumes(unosAlbumes) = self.albumes().addAll(unosAlbumes)	
```
¿Quieren devolver algo? Recuerden que el "=" es equivalente a hacer un cuerpo con un **return**, y nos sirve para ser expresivos y para indicar cuando **no** hay efecto de lado. ¿Pasa esto acá?

sanuzzi commented on src/mundoMusical.wlk:
```diff
@@ -1,41 +1,41 @@
-	}
+	method habilidad() = self.tipoDeMusico().habilidadSegunMusico(self)
+	
+	method habilidad(unaHabilidad) = self.tipoDeMusico().habilidad(unaHabilidad)
```
La habilidad base podría guardarse en el músico.

Y en "habilidad()", el músico le preguntaría al tipoDeMúsico qué habilidad tiene y le enviaría su habilidad base o lo que se necesite.

sanuzzi commented on src/mundoMusical.wlk:
```diff
@@ -47,7 +47,18 @@ class Musico {
+	var habilidad
+	
+	method habilidad() = habilidad
+	method habilidad(unaHabilidad) {
```
La habilidad básica debería tenerla el músico. El tipo de músico (vocalista o de grupo) resolverá la habilidad (real/final) que tiene el músico, en base a si está en grupo o no.

sanuzzi commented on src/mundoMusical.wlk:
```diff
@@ -56,14 +67,22 @@ class TipoMusicoDeGrupo{
+	
+	method habilidadSegunMusico(unaPersona) {
+		if(unaPersona.perteneceUnaBanda()){
+			return unaPersona.tipoDeMusico().habilidad() + self.aumentoDeHabilidad()
```
Lo de "unaPersona.tipoDeMusico().habilidad()" hace ruido, porque en su solución, tipoDeMusico() nos devuelve a nosotros mismos. Es rara esa secuencia.

En cuanto hagan que el músico se guarde su habilidad base, notemos lo siguiente: Lo único que difiere en cada rama, es el "plus" que puede otorgar. Así que podría ser:

method habilidadDe(unMusico) = unMusico.habilidadBase() + self.plusPara(unMusico)

method plusPara(unMusico) = if (unMusico().pertenecesAUnaBanda()) self. aumentoDeHabilidad() else 0

sanuzzi commented on src/mundoMusical.wlk:
```diff
@@ -75,14 +94,20 @@ class TipoMusicoSolista{
 	
-	
+	method habilidadSegunMusico(unaPersona) = if(unaPersona.perteneceUnaBanda()){
+			return unaPersona.tipoDeMusico().habilidad() - 20
```
Idem que lo de músicos de grupo.
Y si se repite, ¿Podremos evitar repetir?

sanuzzi commented on src/mundoMusical.wlk:
```diff
@@ -75,14 +94,20 @@ class TipoMusicoSolista{
 }
 
-class TipoMusicoReCapo {
+class TipoMusicoReCapo inherits TipoDeMusico{
```
Al final esto les va a cambiar en la entrega 4, y lo más sencillo es que luisAlberto sea un WKO, y herede de... ¿Músico? ¿Queremos **todo** lo que hace un músico, o solo una parte? ¿Cómo hacemos para heredar solo lo que en verdad necesitamos que tenga todo músico?

sanuzzi commented on src/presentacion.wlk:
```diff
@@ -11,9 +11,24 @@ class Presentacion {
 	method musicosQueTocan() = musicosQueTocan
 	
-	method tocaSolo(unMusico) =	self.musicosQueTocan().contains(unMusico)
+	method tocaSolo() =	self.musicosQueTocan().size() == 1 // no se si sera la forma mas efefctiva de hacerlo
```
.isEmpty().negate()

sanuzzi commented on src/testEntrega1.wtest:
```diff
@@ -98,12 +84,14 @@ test "Luis Alberto interpreta bien la canci
 test "Joaqu�n act�a solo en La Trastienda y cobra 100" {             //no importa donde act�a, es el precio de la hora
-	assert.equals(joaquin.cobra(laTrastienda), 100)
+	presentacionLaTrastienda.musicosQueTocan().clear()
+	presentacionLaTrastienda.musicosQueTocan().add(joaquin)
```
¡No manoseen la colección desde afuera! Pídanle a la presentación que despida/eche/elimine a los músicos, y luego que agregue a Joaquín.

O bien, sería útil que entienda algo como "musicosQueTocan( **nuevosMúsicos** )", y que eso limpie la colección y la llene de cero. Le enviarían a "[joaquin]", y sería una sola línea.

sanuzzi commented on src/testEntrega2.wtest:
```diff
@@ -76,13 +50,13 @@ test "Cu
 }
 test"Qu� canci�n es la mas larga del �lbum La Sole"{
-	assert.that(laSole.cancionMasLarga() == eres)
+	assert.that(laSole.cancionMasLarga(compararPorLetra) == eres)
```
.equals

sanuzzi commented on src/testEntrega3.wtest:
```diff
@@ -0,0 +1,6 @@
+
+test "testX" {
```
Altos tests

sanuzzi reviewed (commented):
Revisión Entrega "4"

sanuzzi commented on src/Album.wlk:
```diff
@@ -0,0 +1,68 @@
+		fechaLanzamiento = unaFecha
+	}
+
+	method duracionTotalAlbum(){
```
Lo de "Album" no aporta mucho al nombre. Y podrían evitar el return usando el azúcar sintáctico de "method método() = expresión"

sanuzzi commented on src/Album.wlk:
```diff
@@ -0,0 +1,68 @@
+		return self.canciones().sum({cancion => cancion.duracion()})
+	}
+
+	/*method cancionMasLarga(){
```
Si algo queda viejo, bórrenlo o arréglenlo, no dejen código comentado

sanuzzi commented on src/Album.wlk:
```diff
@@ -0,0 +1,68 @@
+
+}
+
+object compararPorLetra {
```
En vez de "comparar..." y "esLaMayor()", podrían ponerle "tamañoSegunLetra", y "tamaño()"

sanuzzi commented on src/canciones.wlk:
```diff
@@ -0,0 +1,28 @@
+	const letra
+	const duracion
+	const titulo
+	const artista
```
¿Le hace falta esto?

Hace que sea medio complicado, porque el artista tiene álbumes, y éstos tienen canciones.

Si asocian en la canción al artista, van a tener que mantener una relación cíclica: Al asignar al artista, tenemos que estar seguros de que también esté en alguno de los álbumes de éste.

No le veo muchas ventajas, evalúen si conviene sacarlo.

sanuzzi commented on src/guitarra.wlk:
```diff
@@ -0,0 +1,16 @@
+class Guitarra {
```
Hagan 2 WKO o 2 clases (si es que quieren muchas Gibson diferentes). Lo importante es que tengan comportamiento diferente. La fender siempre devuelve el mismo valor, y la gibson devuelve un número si está sana y otro en caso contrario.

sanuzzi commented on src/mundoMusical.wlk:
```diff
@@ -0,0 +1,127 @@
+	
+	method aumentoDeHabilidad() = aumentoDeHabilidad
+		
+	method cobra(presentacion) = if (presentacion.tocaSolo()) 100 else 50	
```
El criterio para cobrar ya no está asociado directamente al tipo de músico. Debe ser una entidad separada para combinar como queramos.

sanuzzi commented on src/mundoMusical.wlk:
```diff
@@ -0,0 +1,127 @@
+	
+	method habilidadSegunMusico(unaPersona) {
+		if(unaPersona.perteneceUnaBanda()){
+			return unaPersona.tipoDeMusico().habilidad() + self.aumentoDeHabilidad()
```
¿"unaPersona.tipoDeMusico()" no es lo mismo que "self"?
Igual, la habilidad base debería tenerla el músico.

sanuzzi commented on src/mundoMusical.wlk:
```diff
@@ -0,0 +1,127 @@
+	}
+	method palabra() = palabra
+	
+	method cobra(presentacion) = if (presentacion.lugar().esConcurrido(presentacion.fecha())) 500 else 400
```
Si la presentación sabe tanto el lugar como la fecha... ¿Por qué no puede encargarse ella de decirnos si es concurrida?

Igual esto quizá les cambie un poco con la entrega 4, pero la idea es similar. Deleguen en vez de revolverle las propiedades a otro objeto.

sanuzzi commented on src/mundoMusical.wlk:
```diff
@@ -0,0 +1,127 @@
+		}
+}
+
+class TipoMusicoReCapo inherits TipoDeMusico{
```
Me desdigo respecto a esta recomendación que hice antes: No hagan un tipo de músico re capo.

Hagan un WKO para Luis Alberto, y definan ahí las cosas particulares que él necesita.

Van a notar que comparte "mucho" pero **no "todo"** lo que tienen los músicos normales.

Así que podrían ver cómo heredar solo lo que sí necesitan de un "MusicoAbstracto" o como lo quieran llamar.

De esta forma, tanto Músico como luisAlberto heredarían de esa superclase que solo tenga las cosas en común. ¿Se ve?

sanuzzi commented on src/presentacion.wlk:
```diff
@@ -0,0 +1,34 @@
+	method fecha() = fecha
+	method musicosQueTocan() = musicosQueTocan
+	
+	method tocaSolo() =	self.musicosQueTocan().size() == 1 // no se si sera la forma mas efefctiva de hacerlo
```
Estaría bueno que los músicos/bandas que tocan, estén en un Set.

Una forma un poco más amena de comprobar que la colección solo tiene cierto elemento, es viendo si es equivalente a una nueva colección que solo tenga dicho elemento:

tocaSolo( **unMusico** ) =  self.musicosQueTocan() == #{ **unMusico** }

sanuzzi reviewed (commented):
Faltan los tests de la entrega 3, algunas mejoras, y mucho de la entrega 4.
